File: Client/Data/LocalStore.cs
using Couple.Client.Data.Calendar;
using Couple.Client.Data.ToDo;
using Microsoft.JSInterop;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Couple.Client.Data
{
    public class LocalStore
    {
        protected readonly IJSRuntime js;

        public LocalStore(IJSRuntime js) => (this.js) = (js);

        public ValueTask<T> GetAsync<T>(string storeName, object key)
            => js.InvokeAsync<T>("localStore.get", storeName, key);

        public ValueTask<T> GetAllAsync<T>(string storeName)
            => js.InvokeAsync<T>("localStore.getAll", storeName);

        public ValueTask PutAsync<T>(string storeName, T value)
            => js.InvokeVoidAsync("localStore.put", storeName, value);

        public ValueTask DeleteAsync(string storeName, object key)
            => js.InvokeVoidAsync("localStore.delete", storeName, key);

        // Another possible implementation is to pass only the Event (w/ ToDos),
        // then retrieve the existing Event from database & generate set difference.
        // Pro: Impossible for callers of the method to pass the wrong set of model.
        // Con: Additional code to generate set difference and additional database call.
        // I've used this implementation for code simplicity.
        public ValueTask PutEventAsync(EventModel @event, IEnumerable<Guid> added, IEnumerable<ToDoModel> removed)
            => js.InvokeVoidAsync("localStore.putEvent", @event, added, removed);

        public ValueTask DeleteEventAsync(Guid id, IEnumerable<ToDoModel> removed)
            => js.InvokeVoidAsync("localStore.deleteEvent", id, removed);
    }
}

File: Client/wwwroot/localStore.js
(function () {
    const db = idb.openDB('Couple', 1, {
        upgrade(db) {
            db.createObjectStore('todo', { keyPath: 'id' });
            db.createObjectStore('event', { keyPath: 'id' });
        },
    });

    window.localStore = {
        get: async (storeName, key) => (await db).transaction(storeName).store.get(key),
        getAll: async (storeName) => (await db).transaction(storeName).store.getAll(),
        put: async (storeName, value) => (await db).transaction(storeName, 'readwrite').store.put(value),
        delete: async (storeName, key) => (await db).transaction(storeName, 'readwrite').store.delete(key),

        putEvent: async (event, added, removed) => {
            const tx = (await db).transaction(['event', 'todo'], 'readwrite');
            const eventStore = tx.objectStore('event');
            const toDoStore = tx.objectStore('todo');

            await Promise.all([
                eventStore.put(event),
                added.map(id => toDoStore.delete(id)),
                removed.map(toDo => toDoStore.put(toDo)),
                tx.done,
            ]);
        },

        deleteEvent: async (id, removed) => {
            const tx = (await db).transaction(['event', 'todo'], 'readwrite');
            const eventStore = tx.objectStore('event');
            const toDoStore = tx.objectStore('todo');

            await Promise.all([
                eventStore.delete(id),
                removed.map(toDo => toDoStore.put(toDo)),
                tx.done,
            ]);
        }
    };
})();

